src/Symfony/Component/JsonEncoder/Mapping/PropertyMetadata.php

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\JsonEncoder\Mapping;
+
+use Symfony\Component\JsonEncoder\Exception\InvalidArgumentException;
+use Symfony\Component\TypeInfo\Type;
+
+/**
+ * Holds encoding/decoding metadata about a given property.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental
+ */
+final readonly class PropertyMetadata
+{
+    /**
+     * @param list<\Closure> $formatters
+     */
+    public function __construct(
+        public string $name,
+        public Type $type,
+        public array $formatters = [],
+    ) {
+        self::validateFormatters($this);
+    }
+
+    public function withName(string $name): self
+    {
+        /** @var array{name: string, type: Type, formatters: list<callable>} */
+        $args = (array) $this;
+        $args['name'] = $name;
+
+        return new self(...$args);
```

Comment:
I suggest doing `return new self($name, $this->type, $this->formatters);`, which will be clearer about what actually happens (same in other methods